cunumeric/array.py

```diff
@@ -2017,29 +2023,64 @@ def perform_binary_op(
-                    )
-                    temp._thunk.convert(
-                        one._thunk, stacklevel=(stacklevel + 1)
-                    )
+                    # remove convert ops
+                    if isDeferred and one.shape == ():
+                        temp = ndarray(
+                            shape=one.shape,
+                            dtype=common_type,
+                            # buffer = one._thunk.array.astype(common_type),
+                            stacklevel=(stacklevel + 1),
+                            inputs=(one, two, where),
+                        )
+                        temp._thunk = runtime.create_scalar(
+                            one._thunk.array.astype(common_type),
+                            common_type,
+                            shape=one.shape,
+                            wrap=True,
+                        )
+                    else:
+                        temp = ndarray(
+                            shape=one.shape,
+                            dtype=common_type,
+                            stacklevel=(stacklevel + 1),
+                            inputs=(one, two, where),
+                        )
+                        temp._thunk.convert(
+                            one._thunk, stacklevel=(stacklevel + 1)
+                        )
                     one = temp
                 if two.dtype != common_type:
-                    temp = ndarray(
-                        shape=two.shape,
-                        dtype=common_type,
-                        stacklevel=(stacklevel + 1),
-                        inputs=(one, two, where),
-                    )
-                    temp._thunk.convert(
-                        two._thunk, stacklevel=(stacklevel + 1)
-                    )
+                    # remove convert ops
+                    if isDeferred and two.shape == ():
+                        temp = ndarray(
+                            shape=two.shape,
+                            dtype=common_type,
+                            # buffer = two._thunk.array.astype(common_type),
+                            stacklevel=(stacklevel + 1),
+                            inputs=(one, two, where),
+                        )
+                        temp._thunk = runtime.create_scalar(
+                            two._thunk.array.astype(common_type),
+                            common_type,
+                            shape=two.shape,
+                            wrap=True,
+                        )
+                    else:
+                        temp = ndarray(
+                            shape=two.shape,
+                            dtype=common_type,
+                            stacklevel=(stacklevel + 1),
+                            inputs=(one, two, where),
+                        )
+                        temp._thunk.convert(
+                            two._thunk, stacklevel=(stacklevel + 1)
+                        )
```

Comment:
above code is a scalar constant optimization, which avoids dispatching `CONVERT` operations (for a scalar constant), as the constant's value is embedded in the code and thus already known